frontend/src/components/AppSelect.vue

```diff
@@ -80,81 +85,151 @@ import {
   ListboxOptions,
 } from "@headlessui/vue";
 import AppButton from "@/components/AppButton.vue";
-import { sleep } from "@/util/misc";
+import { frame } from "@/util/misc";
 
-export type Option = { id: string; label: string; colors?: string[] };
+export type Option = {
+  id: string;
+  label: string;
+  [key: string]: unknown;
+};
 
 type Props = {
   label: string;
-  options: Option[];
+  options: O[];
   multi?: boolean;
-  modelValue: Option["id"] | Option["id"][];
+  modelValue: O["id"] | O["id"][];
+  tooltip?: string;
 };
 
-const props = withDefaults(defineProps<Props>(), { multi: false });
+const props = withDefaults(defineProps<Props>(), {
+  multi: false,
+  tooltip: "",
+});
 
 type Emits = {
   "update:modelValue": [Props["modelValue"]];
 };
 
 const emit = defineEmits<Emits>();
 
-// floating-ui middleware
+type Slots = {
+  /** extra preview element to show for each option in dropdown and selected option in label */
+  preview: (props: { option?: O }) => unknown;
+};
+
+defineSlots<Slots>();
+
+/** floating-ui middleware */
 const middleware = [
   size({
     apply({ availableHeight, elements }) {
       Object.assign(elements.floating.style, {
-        // limit popover height to available height
+        /** limit popover height to available height */
         maxHeight: `${availableHeight - 20}px`,
       });
     },
   }),
 ];
 
-// model value to pass from parent to headlessui
-const value = computed(() =>
-  props.multi
-    ? props.options.filter((option) => props.modelValue.includes(option.id))
-    : props.options.find((option) => option.id === props.modelValue) || "",
-);
+/** normalize single/multi to array */
+function normalize<T>(value: T | T[]): T[] {
```

Comment:
Since this component can be in single or multi mode, always normalize (clean) value. Errors I was getting before have now been resolved, but this protects against potential future errors better.

Reply:
I don't love that the function is just called `normalize`. I don't know what I'd call it; `ensureArray` isn't exactly elegant, but that's what came to mind.

Reply:
Will change it to `toArray`.